File: 3020-find-the-maximum-number-of-elements-in-subset/3020-find-the-maximum-number-of-elements-in-subset.cpp
class Solution {
public:
    int maximumLength(vector<int>& nums) {
        int res = 1;
        unordered_map<long, int> ump;
        map<long, int> mp;
        for (auto& x: nums)
            mp[x]++, ump[x]++;
                
        for (auto& [x, freq]: mp) {
            int cnt = 1;
            if (!ump.count(x))
                continue;
            if (x == 1) {
                cnt = (freq % 2) ? freq : freq-1;
            } else {
                long i = x;
                while (ump.count(i) and ump[i] >= 2 and ump.count(i*i)) {
                    cnt += 2;
                    ump[i] -= 2;
                    if (ump[i] <= 0)
                        ump.erase(i);
                    i = i*i;
                }
            }
            res = max(res, cnt);
        }
        return res;
    }
};

File: 3020-find-the-maximum-number-of-elements-in-subset/3020-find-the-maximum-number-of-elements-in-subset.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maximumLength = function(nums) {
    const freq = new Map();

    for (const x of nums) {
        freq.set(x, (freq.get(x) || 0) + 1);
    }

    let ans = 1;

    if (freq.has(1)) {
        const cnt = freq.get(1);

        ans = Math.max(ans, cnt % 2 ? cnt : cnt - 1);
    }

    for (const [start] of freq) {
        if (start === 1) continue;

        let cur = start;
        let len = 0;

        while (freq.has(cur)) {
            if (freq.get(cur) >= 2) {
                len += 2;

                cur = cur * cur;
            } else {
                len++;
                break;
            }
        }

        if (len % 2 === 0) len--;

        ans = Math.max(ans, len);
    }

    return ans;
};